src/audio/copier/copier_hifi.c

```diff
@@ -72,4 +74,67 @@ int apply_attenuation(struct comp_dev *dev, struct copier_data *cd,
 		return -EINVAL;
 	}
 }
+
+void copier_gain_set_basic_params(struct comp_dev *dev, struct dai_data *dd)
+{
+	struct copier_gain_params *gain_params = dd->gain_data;
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+
+	/* Set default gain coefficients */
+	for (int i = 0; i < ARRAY_SIZE(gain_params->gain_coeffs); ++i)
+		gain_params->gain_coeffs[i] = AE_MOVF16X4_FROMINT64(UNITY_GAIN_4X_Q10);
+
+	gain_params->step_f16 = AE_ZERO16();
+	gain_params->init_gain = AE_ZERO16();
+
+	/* Set audio format */
+	gain_params->container = copier_cfg->base.audio_fmt.depth;
+	gain_params->chanels_count = copier_cfg->base.audio_fmt.channels_count;
+}
+
+int copier_gain_set_fade_params(struct comp_dev *dev, struct dai_data *dd,
+				uint32_t fade_period, uint32_t frames)
+{
+	struct copier_gain_params *gain_params = dd->gain_data;
+	struct ipc4_copier_module_cfg *copier_cfg = dd->dai_spec_config;
+	uint16_t init_gain[4];
+	ae_f16 step_f16;
+
+	/* For backward compatibility add a case with default fade transition.
+	 * Backward compatibility is referring to clock_on_delay in DMIC blob.
+	 */
+	if (fade_period == 0) {
+		/* Default fade transition delay for speech */
+		gain_params->fade_sg_length = frames * GAIN_DEFAULT_SPEECH_TRANS_MS;
+
+		/* Default fade transition delay for high quality mode */
+		if (copier_cfg->base.audio_fmt.sampling_frequency > IPC4_FS_16000HZ)
+			gain_params->fade_sg_length = frames * GAIN_DEFAULT_HQ_TRANS_MS;
+
+	} else if (fade_period == GAIN_ZERO_TRANS_MS) {
+		/* Special case for GAIN_ZERO_TRANS_MS to support zero fade in transition time */
+		gain_params->fade_sg_length = 0;
+		return 0;
+	}
+
+	/* High precision step for fade-in calculation, keeps accurate precision */
+	gain_params->step_i64 = MAX_INT64 / gain_params->fade_sg_length;
+	uint16_t step_i64_to_i16 = (uint16_t)(gain_params->step_i64 >> I64_TO_I16_SHIFT);
+
+	step_f16 = step_i64_to_i16 * (MAX_GAIN_COEFFS_CNT / gain_params->chanels_count);
+
+	/* Lower precision step for HIFI SIMD fade-in calculation */
+	gain_params->step_f16 = step_f16;
+
+	/* Initialization gain for HIFI SIMD addition, depends on channel configuration */
+	for (int i = 0; i < MAX_GAIN_COEFFS_CNT; i++)
+		init_gain[i] = (i / gain_params->chanels_count) * step_i64_to_i16;
+
+	int ret = memcpy_s(&gain_params->init_gain, sizeof(gain_params->init_gain), init_gain,
+			   sizeof(init_gain));
+	assert(!ret);
```

Comment:
I hope some day reason will prevail and we'll stop seeing this sort of error handling.